codeU2.py

```diff
@@ -0,0 +1,210 @@
+# -*- coding: utf-8 -*-
+"""
+Created on Fri Jan 13 10:22:28 2017
+Exercise set 2, Code U
+@author: Luiza
+"""
+import logging
+import unittest
+
+def dfs(tree, start):
+
+    """Depth-first search
+    """
+
+    if tree is None:
+        return
+    visited = set()
+    stack = [start]
+    while stack:
+        vertex = stack.pop()
+        if not vertex:
+            continue
+        visited.add(vertex)
+        stack.extend(set(tree[vertex]) - visited)
+    return visited
+
+def dfs_path(tree, start, end, path=None):
+
+    """Returns all full paths from start node
+    to end node, including the end node
+    """
+
+    stack = [(start, [start])]
+    while stack:
+        (vertex, path) = stack.pop()
+        if vertex is None:
+            continue
+        for next in set(tree[vertex]) - set(path):
+            if next == end:
+                yield path + [end]
+            else:
+                stack.append((next, path+[next]))
+
+def get_head(tree):
```

Comment:
Right I have the class for storing all the information about the tree:
class Tree():
    
    """Defines the class for the tree:
    self.children contains a dict of pairs (node:(left_child,right_child))
    self.head contains the value of the tree head
    None denotes the absence if the node
    """
    
    def __init__(self,children,head):
        self.children = children;
        self.head = head;